scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala

```diff
@@ -3286,149 +3472,38 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
-              syntaxError(
-                "extension method cannot have type parameters since some were already given previously",
-                m.tparams.head
-              )
-            }
-          }
-          if (f.is[Defn.Def]) {
-            val m = f.asInstanceOf[Defn.Def]
-            if (m.tparams.nonEmpty) {
-              syntaxError(
-                "extension method cannot have type parameters since some were already given previously",
-                m.tparams.head
-              )
-            }
-          }
-        }
-      }
-    }
-
-    def baseTermDefinedOnce: Unit = {
-      if (eg.eparam.name.value != "") {
-        for (f <- eg.templ.stats) {
-          if (f.is[Defn.ExtensionMethod]) {
-            val term = f.asInstanceOf[Defn.ExtensionMethod].eparam
-            if (term.name.value != "") {
-              syntaxError(
-                "no extension method allowed here since leading parameter was already given",
-                term
-              )
-            }
-          } else if (f.is[Defn.ExtensionMethodInfix]) {
-            val term = f.asInstanceOf[Defn.ExtensionMethodInfix].eparam
-            if (term.name.value != "") {
-              syntaxError(
-                "no extension method allowed here since leading parameter was already given",
-                term
-              )
-            }
-          }
-        }
-      }
-    }
-
-    onlyMethodsOrExtensionMethods
-    baseTermDefinedOnce
-    baseTypeDefinedOnce
-
-    eg
-  }
-
-  // DefDef             ::=  DefSig [(‘:’ | ‘<:’) Type] ‘=’ Expr
-  // DefSig             ::=  ExtParamClause [nl] [‘.’] id DefParamClauses
-  // ExtParamClause     ::=  [DefTypeParamClause] ‘(’ DefParam ‘)’
-  // DefTypeParamClause ::=  ‘[’ DefTypeParam {‘,’ DefTypeParam} ‘]’
-  // DefParamClauses    ::=  {DefParamClause} [[nl] ‘(’ [‘implicit’] DefParams ‘)’]
-  def extMethod(mods: List[Mod]): Stat = {
     val tparams = typeParamClauseOpt(ownerIsType = false, ctxBoundsAllowed = true)
-
-    accept[LeftParen]
-    val basetype =
+    val eparam = inParens(
       param(ownerIsCase = false, ownerIsType = false, isImplicit = false, isUsing = false)
-    accept[RightParen]
-
-    acceptOpt[LF]
-    val isInfix = token.isNot[Dot]
-    acceptOpt[Dot]
-
-    val name = termName()
-    val paramss = paramClauses(ownerIsType = false)
-
-    val decltpe = fromWithinReturnType(typedOpt())
+    )
 
-    accept[Equals]
-    val body = expr()
-    if (isInfix) Defn.ExtensionMethodInfix(mods, basetype, name, tparams, paramss, decltpe, body)
-    else Defn.ExtensionMethod(mods, basetype, name, tparams, paramss, decltpe, body)
+    val methodsAll: List[Stat] = if (token.is[ColonEol]) {
+      accept[ColonEol]
+      indented(templateStats())
+    } else if (token.is[LeftBrace]) {
+      inBraces(templateStats())
+    } else {
+      newLineOpt()
+      List(nonLocalDefOrDcl())
+    }
+    val methods = methodsAll.collect { case m: Defn.Def => m }
```

Comment:
Maybe this should be Block or IndentBlock ?

Reply:
Wrapping it artificially in Block would not give us much here I think.
This is similar to body of class/trait/etc. which is of type 'List[Stat]'.
I can change it to `Template` does that make more sense?
But then Template would be only used for List[Stat] other elements would be unused.

Reply:
i can comment mostly from the standpoint of scalafmt, which needs to know what it is but also how it was specified.

for instance, there are cases (well, like `case` body) where it's parsed as a block but there were no braces, which leads to complicated logic (because expressions defined in a block or a template are considered top level and are occasionally differently formatted; on the other hand, having braces vs not is even more important, because there's a standard indentation rule triggered by braces but when they are not there, it needs to be emulated).

so, in order to format and properly indent this construct, I'd need to identify it as such and also need to know where it starts and ends (and if any leading or trailing comments were also included and need to be indented as well).

that means not a list of statements (they won't include the comments). ideally it'd be a Block enclosed in newly defined tokens LeftImpliedBrace and RightImpliedBrace, which are syntactically empty.

Reply:
This is a good catch, I will try to use 'Term' as body and put there Block / Expr depending on a case.

Artificial tokens would be super hard to add, as we could only know where to add them on parser level but tokenization already happened.

Reply:
body changed to `Stat` that internally can be a `Block` or just single `Def`